chore: local dev setup using minishift

dipak-pawar opened the pull request:
Fixes: #11 

To Do: 
 * [x] Verify all make targets locally(Having some issue with minishift, trying to fix it)

alexeykazakov commented on Dockerfile.dev:
```diff
@@ -1,4 +1,4 @@
-FROM quay.io/openshiftio/fabric8-services-fabric8-cluster:c9398c
+FROM quay.io/openshiftio/fabric8-services-fabric8-cluster:bddfb4e
```
Why not latest?

alexeykazakov replied:
Isn’t it what we want? To use the local build?

alexeykazakov replied:
Isn’t it what we want? To use the local build?

dipak-pawar replied:
for local build we'll be building the latest binary for cluster and just putting it in container using `COPY bin/cluster /usr/local/cluster/bin/cluster`.

alexeykazakov commented on minishift/README.md:
````diff
@@ -0,0 +1,140 @@
+#### Deploying Auth, Cluster together
+To deploy `auth`, `f8cluster` together we have following target:
+```bash
+make deploy-all
````
Shouldn't we start with some quick start instructions in the beginning of the readme? And add all details and additional targets after that?

WDYT about having targets similar to what we have in Auth - https://github.com/fabric8-services/fabric8-auth/blob/master/minishift/README.md
For example `make dev-openshift` to start everything. We can keep your targets too.

alexeykazakov commented on minishift/README.adoc:
````diff
@@ -0,0 +1,206 @@
+make dev-openshift
+```
+Please enter sudo password when prompted, it is needed in order to create an entry in the `/etc/hosts`.
+`minishift ip` gives the IP address on which MiniShift is running. This automation creates a host entry as `minishift.local` for that IP. This domain is whitelisted on fabric8-auth.
````
This domain is whitelisted on fabric8-**cluster**.

dipak-pawar replied:
no, it should `fabric8-auth` because we have whitelisted it in keycloak for auth not for cluster

alexeykazakov commented on minishift/README.adoc:
```diff
@@ -0,0 +1,206 @@
+
+However you can see auth service is running, and will be available at http://auth-fabric8-services.$(minishift_ip).nip.io/api.
+
+If you wish to access the Postgres database, it is available on the same host but on port 31001.  Use the following command to connect with the Postgres client:
```
Duplicates 57-60 but with different port.

dipak-pawar replied:
no one is for auth and one is for cluster service. Let me update it

alexeykazakov commented on minishift/README.adoc:
```diff
@@ -0,0 +1,206 @@
+
+The above command then automates the process of running the containers on OpenShift in MiniShift by using openshift templates.
+
+Once the cluster service is running, it will be available at http://f8cluster-fabric8-services.$(minishift_ip).nip.io/api.
```
I guess we could use minishift.local instead of the IP? It's set via make in /etc/hosts.

dipak-pawar replied:
no we can't.  `nip.io` which is DNS server won't be able to resolve `minishift.local`

alexeykazakov commented:
For cluster pod it's `Container: f8cluster Image Pull Back-off` error 
And I don't understand why the auth pod crashing. The log looks ok:
```
pmcd: cannot open log "/dev/force-logging-to-stderr" for writing : Permission denied
--
  | Log for pmcd on auth-1-8qxsl started Thu Nov 15 22:49:00 2018
  | active agent dom   pid  in out ver protocol parameters
  | ============ === ===== === === === ======== ==========
  | pmcd           2                 2 dso i:6  lib=/var/lib/pcp/pmdas/pmcd/pmda_pmcd.so entry=pmcd_init [0x7fc0c8b82500]
  | proc           3                 2 dso i:7  lib=/var/lib/pcp/pmdas/proc/pmda_proc.so entry=proc_init [0x7fc0c8961620]
  | linux         60                 2 dso i:7  lib=/var/lib/pcp/pmdas/linux/pmda_linux.so entry=linux_init [0x7fc0c872dd90]
  | prometheus   144   121  13  14   2 bin pipe cmd=python /var/lib/pcp/pmdas/prometheus/pmdaprometheus.python
  | Host access list empty: host-based access control turned off
  | User access list empty: user-based access control turned off
  | Group access list empty: group-based access control turned off
  | pmcd: PID = 7, PDU version = 2
  | pmcd request port(s):
  | sts fd   port  family address
  | === ==== ===== ====== =======
  | ok     5       unix   /var/run/pcp/pmcd.socket
  | ok     3 44321 inet   INADDR_ANY
  | ok     4 44321 ipv6   INADDR_ANY
  | [Thu Nov 15 22:49:01] pmcd(7) Info: PMNS file "DEFAULT" is unchanged
  | time="2018-11-15T22:49:04Z" level=warning msg="unable to parse log level configuration error: \"not a valid logrus Level: \\\"\\\"\""
  | time="2018-11-15T22:49:04Z" level=warning msg="Default config is used! This is OK in Dev Mode." default_configuration_error="/etc/fabric8/service-account-secrets.conf is not used; developer Mode is enabled; default service account private key is used; default service account private key ID is used; default user account private key is used; default user account private key ID is used; default DB password is used; default Keycloak client secret is used; default GitHub client secret is used; no restrictions for valid redirect URLs; OSO Reg App url is empty; OSO Reg App admin username is empty; OSO Reg App admin token is empty; environment is expected to be set to 'production' or 'prod-preview'; Sentry DSN is empty"
  | time="2018-11-15 22:49:04" level=warning msg="failed to get current user" err="user: unknown userid 1000160000" file="/tmp/go/src/github.com/fabric8-services/fabric8-auth/main.go" func=main.printUserInfo pkg=main
  | time="2018-11-15 22:49:04" level=info msg="Current version 39. Nothing to update." current_version=39 next_version=40 pkg=migration
  | time="2018-11-15 22:49:04" level=info msg="configuring a new service factory with 0 options" pkg="application/service/factory"
  | time="2018-11-15 22:49:04" level=info msg="public key added" kid=aUGv8mQA85jg4V1DU8Uk1W0uKsxn187KQONAGl6AMtc pkg=token
  | time="2018-11-15 22:49:04" level=info msg="public key added" kid=9MLnViaRkhVj1GT9kpWUkwHIwUD-wZfUxR-3CpkE-Xs pkg=token
  | time="2018-11-15 22:49:04" level=info msg="dev mode public key added" kid=bNq-BCOR3ev-E6buGSaPrU-0SXX8whhDlmZ6geenkTE pkg=token
  | time="2018-11-15 22:49:04" level=warning msg="Tenant service is not enabled"
  | time="2018-11-15 22:49:04" level=info msg="refreshing cached list of clusters..." pkg="cluster/service" req_id=
```
That should be fine if run in DevMode.

alexeykazakov commented:
Auth pod: `Readiness probe failed: Get http://172.17.0.9:8089/api/status: dial tcp 172.17.0.9:8089: connect: connection refused`

alexeykazakov commented:
Cluster pod: `Failed to pull image "fabric8-services/fabric8-cluster:dev": rpc error: code = Unknown desc = repository docker.io/fabric8-services/fabric8-cluster not found: does not exist or no pull access`

dipak-pawar commented:
@alexeykazakov Thanks for trying out locally. This really helped to figure out few issues. I have addressed your comments and  please follow steps to have everything up and running from  https://github.com/fabric8-services/fabric8-cluster/pull/26/files#diff-6d01b92c06052c3ddc0c4eec13d5b11bR43

alexeykazakov commented on minishift/README.adoc:
````diff
@@ -0,0 +1,217 @@
+```
+make minishift-start
+eval $(minishift docker-env)
+make dev-openshift
````
Why not include everything in `make dev-openshift`? Why three commands?

dipak-pawar replied:
If we expose docker-deamon, then it can find locally built docker image sucessfully as we are using `imagePullPolicy: IfNotPresent` which is default in openshift. This way we'll avoid pushing locally built image to minishift registry and pulling it for each deploy which I believe will save lot of our time.

if we use `eval $(minishift docker-env)` inside any make target, all exported variables are only available to that make target's subprocess due to which minishift docker-daemon is not exposed to your host where we built cluster image locally using `docker build -t fabric8-services/fabric8-cluster:dev -f Dockerfile.dev .` To avoid we need to use three targets.
